snapcraft/internal/lifecycle/_runner.py

```diff
@@ -316,6 +331,28 @@ def _handle_dirty(self, part, step, dirty_report, cli_config):
         getattr(self, '_re{}'.format(step.name))(part, hint='({})'.format(
             dirty_report.get_summary()))
 
+    def _handle_outdated(self, part, step, outdated_report, cli_config):
```

Comment:
same as handle_dirty, can this take a `dirty_action` as a parameter instead of passing down the entire object?

Reply:
With the introduction of the outdated concept, I think this is just another case of dirty. Can we add more precise qualifiers like `handle_dirty_definitions` and `handle_dirty_[local]_source`?

Reply:
Let me give you my thought process. A step (for a specific part) can be in one of two states:

1. Not run (we call this "clean")
2. Run, which is further broken down into sub-states:
   2.1. Complete
   2.2. Needs to be cleaned and run again (we call this "dirty", i.e. it needs to be cleaned)
   2.3. Needs to be run again, but does not need to be cleaned (in this PR we call this "outdated", i.e. it needs to be updated)

Both 2.2 and 2.3 are _caused_ by something, which needs to be recorded somewhere so we can tell the user what it was. We do that with the `{Dirty,Outdated}Report` classes. This allows us to isolate concerns, separating the "what makes this {dirty,outdated}" from "oh, this is {dirty,outdated}, I need to take appropriate action." If we combine these reports into one, we have a few issues:

1. We would need to come up with a new name, since "dirty" no longer works (i.e. the solution is not to clean it anymore).
2. Concerns leak. Now lifecycle needs to know "Oh, the project properties changed, okay, I know I need to clean" or "Oh, an earlier step happened, I know I need to update". This spread of knowledge seems error-prone.

(1) isn't a big deal, we can figure that one out. The word "outdated" could be used to refer to both situations, I suppose. (2) we can probably solve by keeping things similar to how they are today, but have a master class composed of both a dirty and outdated report, and take action appropriately. That at least gives us a single report to deal with in the `PluginHandler`'s API and thus the lifecycle, but other than that doesn't buy us much. Do you like that idea better?

Reply:
I did not say combine them, I just meant that "dirty" and "outdated" are easier to confuse and that we should narrow the scope of what it means down.

Your thought process is from the point of view of the action wrt lifecycle and my mind is trending towards the state of things (why is it dirty/outdated) whilst the report comes from a component that does not determine the future actions.

These two terms lead to confusions, is this dirty because it is outdated (so far that has been our logic, right?). The _what_ is outdated is what I am asking to more clearly state in the method calls, checks and class names.

But yeah, by no means, unless it gets converted into the state class merge these two.

Reply:
If you are really stuck with the names, I have added a couple more comments that would satisfy me. My intention was that from the names of variables and classes it was crystal clear what it meant without having to read the code, I can compromise with a bit more documentation under the _internal_ class names and methods called.

Reply:
I agree that this should be cleaned up, but as we discussed on the call, let's see if we can get this PR in a decent state with docs and then take another pass (in another PR) at improving it further.